File: 入會任務/Chinese Chess (象棋)/src/main/java/com/example/ChineseChess.java
package com.example;

import java.util.Map;

public class ChineseChess {

    private Map<String, String> boardState;
    private boolean gameOver = false;

    public ChineseChess() {
        this.boardState = new java.util.HashMap<>();
    }

    public void setBoard(Map<String, String> boardState) {
        this.boardState = new java.util.HashMap<>(boardState);
        this.gameOver = false;
    }

    public boolean isLegalMove(String piece, int fromRow, int fromCol, int toRow, int toCol) {
        boolean isLegal = false;
        switch (piece) {
            case "Red General":
                isLegal = isLegalGeneralMove(fromRow, fromCol, toRow, toCol);
                break;
            case "Red Guard":
                isLegal = isLegalGuardMove(fromRow, fromCol, toRow, toCol);
                break;
            case "Red Rook":
                isLegal = isLegalRookMove(fromRow, fromCol, toRow, toCol);
                break;
            case "Red Horse":
                isLegal = isLegalHorseMove(fromRow, fromCol, toRow, toCol);
                break;
            case "Red Cannon":
                isLegal = isLegalCannonMove(fromRow, fromCol, toRow, toCol);
                break;
            case "Red Elephant":
                isLegal = isLegalElephantMove(fromRow, fromCol, toRow, toCol);
                break;
            case "Red Soldier":
                isLegal = isLegalSoldierMove(fromRow, fromCol, toRow, toCol);
                break;
        }

        if (isLegal) {
            String targetPosition = String.format("(%d, %d)", toRow, toCol);
            if (boardState.containsValue(targetPosition)) {
                String capturedPiece = "";
                for (Map.Entry<String, String> entry : boardState.entrySet()) {
                    if (entry.getValue().equals(targetPosition)) {
                        capturedPiece = entry.getKey();
                        break;
                    }
                }
                if ("Black General".equals(capturedPiece)) {
                    gameOver = true;
                }
                boardState.remove(capturedPiece);
            }
        }

        return isLegal;
    }

    private boolean isLegalSoldierMove(int fromRow, int fromCol, int toRow, int toCol) {
        // Before crossing the river (rows 1-5 for Red)
        if (fromRow <= 5) {
            return toRow == fromRow + 1 && toCol == fromCol;
        } else { // After crossing the river
            return (toRow == fromRow + 1 && toCol == fromCol) || (Math.abs(toCol - fromCol) == 1 && toRow == fromRow);
        }
    }

    private boolean isLegalElephantMove(int fromRow, int fromCol, int toRow, int toCol) {
        int rowDiff = Math.abs(toRow - fromRow);
        int colDiff = Math.abs(toCol - fromCol);

        if (rowDiff != 2 || colDiff != 2) {
            return false;
        }

        // Cannot cross the river
        if (toRow > 5) {
            return false;
        }

        // Check for blocking piece at the midpoint
        if (isPieceAt(fromRow + (toRow - fromRow) / 2, fromCol + (toCol - fromCol) / 2)) {
            return false;
        }

        return true;
    }

    private boolean isLegalCannonMove(int fromRow, int fromCol, int toRow, int toCol) {
        if (fromRow != toRow && fromCol != toCol) {
            return false;
        }

        // Check for pieces in the path
        int piecesInPath = 0;
        if (fromRow == toRow) { // Horizontal move
            int start = Math.min(fromCol, toCol) + 1;
            int end = Math.max(fromCol, toCol);
            for (int c = start; c < end; c++) {
                if (isPieceAt(fromRow, c)) {
                    piecesInPath++;
                }
            }
        } else { // Vertical move
            int start = Math.min(fromRow, toRow) + 1;
            int end = Math.max(fromRow, toRow);
            for (int r = start; r < end; r++) {
                if (isPieceAt(r, fromCol)) {
                    piecesInPath++;
                }
            }
        }

        if (isPieceAt(toRow, toCol)) { // Capture move
            return piecesInPath == 1;
        } else { // Non-capture move
            return piecesInPath == 0;
        }
    }

    private boolean isLegalHorseMove(int fromRow, int fromCol, int toRow, int toCol) {
        int rowDiff = Math.abs(toRow - fromRow);
        int colDiff = Math.abs(toCol - fromCol);

        if (!((rowDiff == 2 && colDiff == 1) || (rowDiff == 1 && colDiff == 2))) {
            return false;
        }

        // Check for blocking piece
        if (rowDiff == 2) { // Moving 2 rows
            if (isPieceAt(fromRow + (toRow - fromRow) / 2, fromCol)) {
                return false;
            }
        } else { // Moving 2 cols
            if (isPieceAt(fromRow, fromCol + (toCol - fromCol) / 2)) {
                return false;
            }
        }

        return true;
    }

    private boolean isLegalRookMove(int fromRow, int fromCol, int toRow, int toCol) {
        if (fromRow != toRow && fromCol != toCol) {
            return false;
        }

        // Check for pieces in the path
        if (fromRow == toRow) { // Horizontal move
            int start = Math.min(fromCol, toCol) + 1;
            int end = Math.max(fromCol, toCol);
            for (int c = start; c < end; c++) {
                if (isPieceAt(fromRow, c)) {
                    return false;
                }
            }
        } else { // Vertical move
            int start = Math.min(fromRow, toRow) + 1;
            int end = Math.max(fromRow, toRow);
            for (int r = start; r < end; r++) {
                if (isPieceAt(r, fromCol)) {
                    return false;
                }
            }
        }

        return true;
    }

    private boolean isPieceAt(int row, int col) {
        if (boardState == null) {
            return false;
        }
        String position = String.format("(%d, %d)", row, col);
        return boardState.containsValue(position);
    }

    private boolean isLegalGuardMove(int fromRow, int fromCol, int toRow, int toCol) {
        if (!isInsideRedPalace(toRow, toCol)) {
            return false;
        }

        int rowDiff = Math.abs(toRow - fromRow);
        int colDiff = Math.abs(toCol - fromCol);
        return rowDiff == 1 && colDiff == 1;
    }

    private boolean isLegalGeneralMove(int fromRow, int fromCol, int toRow, int toCol) {
        if (!isInsideRedPalace(toRow, toCol)) {
            return false;
        }

        // Generals facing each other rule
        if (boardState != null && boardState.containsKey("Black General")) {
            int[] blackGeneralPosition = parsePosition(boardState.get("Black General"));
            int blackGeneralCol = blackGeneralPosition[1];
            if (toCol == blackGeneralCol) {
                return false; // Illegal move if they are on the same file
            }
        }

        int rowDiff = Math.abs(toRow - fromRow);
        int colDiff = Math.abs(toCol - fromCol);
        return (rowDiff == 1 && colDiff == 0) || (rowDiff == 0 && colDiff == 1);
    }

    private int[] parsePosition(String position) {
        String[] parts = position.replaceAll("[()]", "").split(",");
        return new int[]{Integer.parseInt(parts[0].trim()), Integer.parseInt(parts[1].trim())};
    }

    private boolean isInsideRedPalace(int row, int col) {
        // Palace for Red is rows 1-3, cols 4-6
        return row >= 1 && row <= 3 && col >= 4 && col <= 6;
    }

    public boolean isGameOver() {
        return gameOver;
    }
}

File: 入會任務/Chinese Chess (象棋)_Dart/lib/src/chinese_chess.dart
import 'dart:math';
import 'board.dart';
import 'piece.dart';

class ChineseChess {
  final Board board = Board();
  GameStatus gameStatus = GameStatus.ongoing;

  bool isLegalMove(int fromRow, int fromCol, int toRow, int toCol) {
    final piece = board.getPiece(fromRow, fromCol);
    if (piece == null) {
      return false;
    }

    if (!_isPieceMovementLegal(piece, fromRow, fromCol, toRow, toCol)) {
      return false;
    }

    return _isMoveLegalUnderGeneralRules(fromRow, fromCol, toRow, toCol);
  }

  void move(int fromRow, int fromCol, int toRow, int toCol) {
    final targetPiece = board.getPiece(toRow, toCol);
    if (targetPiece?.type == PieceType.General) {
      gameStatus = GameStatus.red_wins; // Assuming Red is the one capturing
    }
    board.movePiece(fromRow, fromCol, toRow, toCol);
  }

  bool _isPieceMovementLegal(Piece piece, int fromRow, int fromCol, int toRow, int toCol) {
    switch (piece.type) {
      case PieceType.General:
        return _isLegalGeneralMove(fromRow, fromCol, toRow, toCol, piece.color);
      case PieceType.Guard:
        return _isLegalGuardMove(fromRow, fromCol, toRow, toCol, piece.color);
      case PieceType.Elephant:
        return _isLegalElephantMove(fromRow, fromCol, toRow, toCol, piece.color);
      case PieceType.Horse:
        return _isLegalHorseMove(fromRow, fromCol, toRow, toCol);
      case PieceType.Rook:
        return _isLegalRookMove(fromRow, fromCol, toRow, toCol);
      case PieceType.Cannon:
        return _isLegalCannonMove(fromRow, fromCol, toRow, toCol);
      case PieceType.Soldier:
        return _isLegalSoldierMove(fromRow, fromCol, toRow, toCol, piece.color);
      default:
        return false;
    }
  }

  bool _isMoveLegalUnderGeneralRules(int fromRow, int fromCol, int toRow, int toCol) {
    final pieceToMove = board.getPiece(fromRow, fromCol)!;
    final pieceAtTarget = board.getPiece(toRow, toCol);

    board.movePiece(fromRow, fromCol, toRow, toCol);

    bool isIllegal = _isGeneralsFacingEachOther();

    board.movePiece(toRow, toCol, fromRow, fromCol);
    if (pieceAtTarget != null) {
      board.addPiece(pieceAtTarget, toRow, toCol);
    } else {
      board.removePiece(fromRow, fromCol);
      board.addPiece(pieceToMove, fromRow, fromCol);
    }

    return !isIllegal;
  }

  bool _isLegalGeneralMove(int fromRow, int fromCol, int toRow, int toCol, PlayerColor color) {
    if (!_isInPalace(toRow, toCol, color)) {
      return false;
    }
    final rowDiff = (toRow - fromRow).abs();
    final colDiff = (toCol - fromCol).abs();
    return rowDiff + colDiff == 1;
  }

  bool _isLegalGuardMove(int fromRow, int fromCol, int toRow, int toCol, PlayerColor color) {
    if (!_isInPalace(toRow, toCol, color)) {
      return false;
    }
    final rowDiff = (toRow - fromRow).abs();
    final colDiff = (toCol - fromCol).abs();
    return rowDiff == 1 && colDiff == 1;
  }

  bool _isLegalElephantMove(int fromRow, int fromCol, int toRow, int toCol, PlayerColor color) {
    if (color == PlayerColor.Red && toRow > 5) return false;
    if (color == PlayerColor.Black && toRow < 6) return false;

    final rowDiff = (toRow - fromRow).abs();
    final colDiff = (toCol - fromCol).abs();
    if (rowDiff != 2 || colDiff != 2) return false;

    final midRow = (fromRow + toRow) ~/ 2;
    final midCol = (fromCol + toCol) ~/ 2;
    return board.getPiece(midRow, midCol) == null;
  }

  bool _isLegalHorseMove(int fromRow, int fromCol, int toRow, int toCol) {
    final rowDiff = (toRow - fromRow).abs();
    final colDiff = (toCol - fromCol).abs();
    if (!((rowDiff == 2 && colDiff == 1) || (rowDiff == 1 && colDiff == 2))) {
      return false;
    }

    if (rowDiff == 2) {
      if (board.getPiece(fromRow + (toRow - fromRow) ~/ 2, fromCol) != null) {
        return false;
      }
    } else { // colDiff == 2
      if (board.getPiece(fromRow, fromCol + (toCol - fromCol) ~/ 2) != null) {
        return false;
      }
    }
    return true;
  }

  bool _isLegalRookMove(int fromRow, int fromCol, int toRow, int toCol) {
    return _countPiecesBetween(fromRow, fromCol, toRow, toCol) == 0;
  }

  bool _isLegalCannonMove(int fromRow, int fromCol, int toRow, int toCol) {
    final piecesBetween = _countPiecesBetween(fromRow, fromCol, toRow, toCol);
    final targetPiece = board.getPiece(toRow, toCol);

    if (targetPiece == null) { // Move
      return piecesBetween == 0;
    } else { // Capture
      return piecesBetween == 1;
    }
  }

  bool _isLegalSoldierMove(int fromRow, int fromCol, int toRow, int toCol, PlayerColor color) {
    final rowDiff = toRow - fromRow;
    final colDiff = (toCol - fromCol).abs();

    if (color == PlayerColor.Red) {
      if (fromRow <= 5) { // Before crossing river
        return rowDiff == 1 && colDiff == 0;
      } else { // After crossing river
        return (rowDiff == 1 && colDiff == 0) || (rowDiff == 0 && colDiff == 1);
      }
    } else { // Black
      if (fromRow >= 6) { // Before crossing river
        return rowDiff == -1 && colDiff == 0;
      } else { // After crossing river
        return (rowDiff == -1 && colDiff == 0) || (rowDiff == 0 && colDiff == 1);
      }
    }
  }

  bool _isInPalace(int row, int col, PlayerColor color) {
    if (color == PlayerColor.Red) {
      return row >= 1 && row <= 3 && col >= 4 && col <= 6;
    } else {
      return row >= 8 && row <= 10 && col >= 4 && col <= 6;
    }
  }

  bool _isGeneralsFacingEachOther() {
    final redPos = board.redGeneralPos;
    final blackPos = board.blackGeneralPos;
    if (redPos == null || blackPos == null) return false;
    if (redPos[1] != blackPos[1]) return false;
    return _countPiecesBetween(redPos[0], redPos[1], blackPos[0], blackPos[1]) == 0;
  }

  int _countPiecesBetween(int r1, int c1, int r2, int c2) {
    if (r1 != r2 && c1 != c2) return -1; // Not a straight line
    int count = 0;
    if (r1 == r2) {
      for (int c = min(c1, c2) + 1; c < max(c1, c2); c++) {
        if (board.getPiece(r1, c) != null) count++;
      }
    } else {
      for (int r = min(r1, r2) + 1; r < max(r1, r2); r++) {
        if (board.getPiece(r, c1) != null) count++;
      }
    }
    return count;
  }
}

enum GameStatus { ongoing, red_wins, black_wins }